ADAL/src/urlprotocol/ADURLProtocol.m

```diff
@@ -275,35 +247,87 @@ - (NSURLRequest *)connection:(NSURLConnection *)connection
     // The following ends up calling -URLSession:task:didCompleteWithError: with NSURLErrorDomain / NSURLErrorCancelled,
     // which specificallys traps and ignores the error.
     
-    [_connection cancel];
+    [task cancel];
     [self.client URLProtocol:self
             didFailWithError:[NSError errorWithDomain:NSCocoaErrorDomain
                                                  code:NSUserCancelledError
                                              userInfo:nil]];
     
-    return mutableRequest;
+    completionHandler(mutableRequest);
 }
 
-- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task
+didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
+ completionHandler:(ChallengeCompletionHandler)completionHandler
 {
-    (void)connection;
+    NSString *authMethod = [challenge.protectionSpace.authenticationMethod lowercaseString];
+    AD_LOG_VERBOSE_F(@"session:task:didReceiveChallenge:completionHandler", _correlationId,
+                     @"%@. Previous challenge failure count: %ld", authMethod, (long)challenge.previousFailureCount);
     
-    [self.client URLProtocol:self didReceiveResponse:response cacheStoragePolicy:NSURLCacheStorageNotAllowed];
+    BOOL handled = NO;
+    Class<ADAuthMethodHandler> handler = nil;
+    @synchronized ([self class]) {
+        handler = [s_handlers objectForKey:authMethod];
+    }
+    
+    handled = [handler handleChallenge:challenge
+                               session:session
+                                  task:task
+                              protocol:self
+                     completionHandler:completionHandler];
+    
+    if (!handled)
+    {
+        // Do default handling
+        completionHandler(NSURLSessionAuthChallengePerformDefaultHandling, nil);
+        return;
+    }
+    
+    if ([authMethod caseInsensitiveCompare:NSURLAuthenticationMethodNTLM] == NSOrderedSame)
+    {
+        [s_telemetryEvent setNtlm:AD_TELEMETRY_YES];
+    }
 }
 
-- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(nullable NSError *)error
 {
-    (void)connection;
+    (void)session;
+    (void)task;
     
-    [self.client URLProtocol:self didLoadData:data];
+    if (error == nil)
+    {
+        [self.client URLProtocolDidFinishLoading:self];
+    }
+    else if ([error.domain isEqual:NSURLErrorDomain] && error.code == NSURLErrorCancelled)
+    {
+        // Do nothing.  This happens in two cases - 
+        // during a redirect, and request cancellation via -stopLoading.
+    }
+    else
+    {
+        [self.client URLProtocol:self didFailWithError:error];
```

Comment:
There is already error logging when they are created or processed by the underlying code, adding extra logging here will result in duplicative logging.